File: 238-product-of-arr-except-self/solution.go
func initProdArr(size int) []int {
	arr := make([]int, size)
	for idx := range size {
		arr[idx] = 1
	}
	return arr
}
func productExceptSelf(nums []int) []int {
	size := len(nums)
	leftProds := initProdArr(size)
	rightProds := initProdArr(size)

	for idx := 1; idx < size; idx++ {
		leftProds[idx] = nums[idx-1] * leftProds[idx-1]
	}

	for idx := size - 2; idx > -1; idx-- {
		rightProds[idx] = nums[idx+1] * rightProds[idx+1]
	}

	for idx := range size {
		nums[idx] = leftProds[idx] * rightProds[idx]
	}

	return nums
}

File: 238-product-of-arr-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        leftProd = [1] * n
        rightProd = [1] * n

        # sum[i:j] = sum[j] - sum[i-1]

        for i in range(1, n):
            leftProd[i] = nums[i-1] * leftProd[i-1]
        
        for i in range(n-2, -1, -1):
            rightProd[i] = nums[i+1] * rightProd[i+1]

        return [leftProd[i] * rightProd[i] for i in range(n)]
        